src/test/fio/fio_ceph_messenger.cc

```diff
@@ -0,0 +1,690 @@
+  unsigned int io_pending_nr  = 0;
+};
+
+struct ceph_msgr_io {
+  struct flist_head list;
+  struct ceph_msgr_data *data;
+  struct io_u *io_u;
+  MOSDOp *req_msg; /** Cached request, valid only for sender */
+};
+
+struct ceph_msgr_reply_io {
+  struct flist_head list;
+  MOSDOpReply *rep;
+};
+
+static void *str_to_ptr(const std::string &str)
+{
+  return (void *)strtoul(str.c_str(), NULL, 16);
+}
+
+static std::string ptr_to_str(void *ptr)
+{
+  char buf[32];
+
+  snprintf(buf, sizeof(buf), "%llx", (unsigned long long)ptr);
+  return std::string(buf);
+}
+
+/*
+ * Used for refcounters print on the last context put, almost duplicates
+ * global context refcounter, sigh.
+ */
+static std::atomic<int> ctx_ref(1);
+
+static void create_or_get_ceph_context(struct ceph_msgr_options *o)
+{
+  if (g_ceph_context) {
+    g_ceph_context->get();
+    ctx_ref++;
+    return;
+  }
+
+  boost::intrusive_ptr<CephContext> cct;
+  vector<const char*> args;
+
+  if (o->conffile)
+    args = { "--conf", o->conffile };
+
+  cct = global_init(NULL, args, CEPH_ENTITY_TYPE_CLIENT,
+		    CODE_ENVIRONMENT_UTILITY,
+		    CINIT_FLAG_NO_DEFAULT_CONFIG_FILE);
+  /* Will use g_ceph_context instead */
+  cct.detach();
+
+  common_init_finish(g_ceph_context);
+  g_ceph_context->_conf.apply_changes(NULL);
+}
+
+static void put_ceph_context(void)
+{
+  if (--ctx_ref == 0) {
+    ostringstream ostr;
+    Formatter* f;
+
+    f = Formatter::create("json-pretty");
+    g_ceph_context->get_perfcounters_collection()->dump_formatted(f, false);
+    ostr << ">>>>>>>>>>>>> PERFCOUNTERS BEGIN <<<<<<<<<<<<" << std::endl;
+    f->flush(ostr);
+    ostr << ">>>>>>>>>>>>>  PERFCOUNTERS END  <<<<<<<<<<<<" << std::endl;
+
+    delete f;
+    dout(0) <<  ostr.str() << dendl;
+  }
+
+  g_ceph_context->put();
+}
+
+static void ceph_msgr_sender_on_reply(const object_t &oid)
+{
+  struct ceph_msgr_data *data;
+  struct ceph_msgr_io *io;
+
+  /*
+   * Here we abuse object and use it as a raw pointer. Since this is
+   * only for benchmarks and testing we do not care about anything
+   * but performance.  So no need to use global structure in order
+   * to search for reply, just send a pointer and get it back.
+   */
+
+  io = (decltype(io))str_to_ptr(oid.name);
+  data = io->data;
+  ring_buffer_enqueue(&data->io_completed_q, (void *)io);
+}
+
+
+class ReplyCompletion : public Message::CompletionHook {
+  struct ceph_msgr_io *m_io;
+
+public:
+  ReplyCompletion(MOSDOpReply *rep, struct ceph_msgr_io *io) :
+    Message::CompletionHook(rep),
+    m_io(io) {
+  }
+  void finish(int err) override {
+    struct ceph_msgr_data *data = m_io->data;
+
+    ring_buffer_enqueue(&data->io_completed_q, (void *)m_io);
+  }
+};
+
+static void ceph_msgr_receiver_on_request(struct ceph_msgr_data *data,
+					  MOSDOp *req)
+{
+  MOSDOpReply *rep;
+
+  rep = new MOSDOpReply(req, 0, 0, 0, false);
+  rep->set_connection(req->get_connection());
+
+  pthread_spin_lock(&data->spin);
```

Comment:
> this lock is a bottleneck for server. my test 10GNIC +4k read, only 13K.

Nope, it is not. To avoid any backpressure iodepth= on server side should be equal or bigger, than on client side, what is yours? 

The following is the result on localhost, default config (bs=4k, iodepth=128 for client and server):
IOPS=112k, BW=439MiB/s

Reply:
@rouming . I'll try your method.  But i change the code as i putted, the performance improved from 13K to 102K.

Reply:
@rouming . By your suggestion, the performance is the same as with my pr. But we i improve numjobs > 1, the server crushed.  For this bug are you met?

Reply:
Well, in order not to sacrifice performance completion queue is a lockless ring buffer for 1 consumer and 1 producer. When you increase number of clients, server starts consuming fio requests from two threads (two client connections), i.e. ring buffer should be modified in order to support N producers. After some time playing with messenger I found that it is not very much interesting to test multiple connections - performance is bad even for single client, so in order to analyse where we spend most of the time I decided to switch to lockless ring buffer and to make 1 connection case as fast as possible. 

If you still want to play with N clients - 1 server, here it is:

```
diff --git a/src/test/fio/fio_ceph_messenger.cc b/src/test/fio/fio_ceph_messenger.cc
index f631387f6e7b..0f858f774467 100644
--- a/src/test/fio/fio_ceph_messenger.cc
+++ b/src/test/fio/fio_ceph_messenger.cc
@@ -151,7 +151,9 @@ static void ceph_msgr_sender_on_reply(const object_t &oid)
 
   io = (decltype(io))str_to_ptr(oid.name);
   data = io->data;
+  pthread_spin_lock(&data->spin);
   ring_buffer_enqueue(&data->io_completed_q, (void *)io);
+  pthread_spin_unlock(&data->spin);
 }
 
 
@@ -166,7 +168,9 @@ public:
   void finish(int err) override {
     struct ceph_msgr_data *data = m_io->data;
 
+    pthread_spin_lock(&data->spin);
     ring_buffer_enqueue(&data->io_completed_q, (void *)m_io);
+    pthread_spin_unlock(&data->spin);
   }
 };
 
@@ -527,7 +531,9 @@ static struct io_u *fio_ceph_msgr_event(struct thread_data *td, int event)
   struct ceph_msgr_io *io;
 
   data = (decltype(data))td->io_ops_data;
+  pthread_spin_lock(&data->spin);
   io = (decltype(io))ring_buffer_dequeue(&data->io_completed_q);
+  pthread_spin_unlock(&data->spin);
 
   return io->io_u;
 }
```